README.md

```diff
@@ -1,10 +1,238 @@
+
+## Implementation
+
+The SMT has 4 node types that are used to construct the tree:
+
+- Inner Nodes
+  - Prefixed `[]byte{1}`
+  - `digest = hash([]byte{1} + leftChild.digest + rightChild.digest)`
+- Extension Nodes
+  - Prefixed `[]byte{2}`
+  - `digest = hash([]byte{2} + pathBounds + path + child.digest)`
+- Leaf Nodes
+  - Prefixed `[]byte{0}`
+  - `digest = hash([]byte{0} + path + value)`
+- Lazy Nodes
+  - Prefix of the actual node type is stored in the digest
+  - `digest = persistedDigest`
+
+### Inner Nodes
+
+Inner nodes represent a branch in the tree with two **non-nil** child nodes. The inner node has an internal `digest` which represents the hash of the child nodes concatenated hashes.
+
+### Extension Nodes
+
+Extension nodes represent a singly linked chain of inner nodes, with a single child. They are used to represent a common path in the tree and as such contain the path and bounds of the path they represent. The `digest` of an extension node is the hash of its path bounds, the path itself and the child nodes digest concatenated.
+
+### Leaf Nodes
+
+Leaf nodes store the full path which they represent and also the hash of the value they store. The `digest` of a leaf node is the hash of the leaf nodes path and value concatenated.
+
+The SMT stores only the hashes of the values in the tree, not the raw values themselves. In order to store the raw values in the underlying database the option `WithValueHasher(nil)` must be passed into the `NewSparseMerkleTree` constructor.
+
+### Lazy Nodes
+
+Lazy nodes represent uncached, persisted nodes, and as such only store the `digest` of the node. When a lazy node is accessed the node type will be determined and the full node type will be populated with any relevant fields such as its children and path.
+
+### Lazy Loading
+
+This library uses a cached, lazy-loaded tree structure to optimize performance. It optimises performance by not reading from/writing to the underlying database on each operation, deferring any underlying changes until the `Commit()` function is called.
+
+All nodes have a `persisted` field which signals whether they have been persisted to the underlying database or not. In practice this gives a large performance optimisation by working on cached data and not reading from/writing to the database on each operation. If a node is deleted from the tree it is marked as `orphaned` and will be deleted from the database when the `Commit()` function is called.
+
+Once the `Commit()` function is called the tree will delete any orphaned nodes from the database and write the key-value pairs of all the unpersisted leaf nodes' hashes and their values to the database.
+
+### Visualisations
+
+The following diagrams are representations of how the tree and its components can be visualised.
+
+#### General Tree Structure
+
+The different nodes types described above make the tree have a structure similar to the following:
+
+```mermaid
```

Comment:
I think you made really good progress on this, but there is still a lot of opportunity to improve it so it's easier for the reader to understand & interpret. Similar to what the JMT and Relay Mining Paper do. For example, by showing `(k, v) = (a, b)` or `path=0b001` or `path_bounds=...`. 

In either case, the visualizer we build might do this automatically for us in the future (we can just have a python structure to mermaid converter) so not a blocker for now.


![Screenshot 2023-05-31 at 6 29 02 PM](https://github.com/pokt-network/smt/assets/1892194/a5a68d85-ef77-4452-b25c-2c47424322ca)

![Screenshot 2023-05-31 at 6 25 14 PM](https://github.com/pokt-network/smt/assets/1892194/d02830fe-0048-4c99-9d8e-9dc2d34933e9)

https://developers.diem.com/papers/jellyfish-merkle-tree/2021-01-14.pdf
https://arxiv.org/pdf/2305.10672.pdf

Reply:
I have added an interim diagram before we can have a more detailed one to show the paths of different nodes